src/import-shadow-hash.c

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright 2016 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "storage.h"
+#include "module.h"
+#include "entry.h"
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <shadow.h>
+
+/**
+ * CreateEntryFromShadowHash:
+ * @username: Username containing the shadow hash.
+ * @sdhash: User's hashed shadow password.
+ */
+gboolean CreateEntryFromShadowHash(const char *username, char *sdhash) {
```

Comment:
Put almost all of the logic of this file into this function, and unit test it. Make an argument to this with each flag that you create for main. Also can you rename it to:

`gboolean CacheImport(const char *shadow_path, const char *storage_path, const char *username, GError **error)`

If lock/unlock isn't in main() anymore, then all main() does is call this with flag/argv values and check/print any error.